runtime/src/tasks/qthreads/tasks-qthreads.c

```diff
@@ -657,9 +657,49 @@ static void setupAffinity(void) {
     chpl_qt_setenv("AFFINITY", "no", 0);
   }
 
-  // For the binders topo spread threads across sockets instead of packing.
-  // Only impacts binders, but it doesn't hurt to set it for other configs.
-  chpl_qt_setenv("LAYOUT", "BALANCED", 0);
+  // Explicitly bind shepherd threads to cores/PUs when using the
+  // binders topology.
+  if (CHPL_QTHREAD_TOPOLOGY_BINDERS) {
+    int *cpus = NULL;
+    int numCpus;
+    chpl_bool physical;
+    char *unit = chpl_qt_getenv_str("WORKER_UNIT");
+    if ((unit != NULL) && !strcmp(unit, "pu")) {
+        physical = false;
+        numCpus = chpl_topo_getNumCPUsLogical(true);
+    } else {
+        physical = true;
+        numCpus = chpl_topo_getNumCPUsPhysical(true);
+    }
+    int numShepherds = (int) chpl_qt_getenv_num("NUM_SHEPHERDS", numCpus);
+    if (numShepherds < numCpus) {
+        numCpus = numShepherds;
+    }
+    cpus = (int *) chpl_malloc(numCpus * sizeof(*cpus));
+    numCpus = chpl_topo_getCPUs(physical, cpus, numCpus);
+    if (numCpus > 0) {
+      // Determine how much space we need for the string of CPU IDs.
+      // Note: last ':' will be replaced with NULL.
+      int bufSize = 0;
+      for (int i = 0; i < numCpus; i++) {
+        bufSize +=  snprintf(NULL, 0, "%d:", cpus[i]);
```

Comment:
Clever. I wasn't sure if this was standard, but https://stackoverflow.com/questions/35035982/is-snprintfnull-0-behavior-standardized seems to indicate yes.

Reply:
Yes, I've done it this way before.